// @graphql-codegen/visitor-plugin-common/dist/esnext/client-side-base-visitor.js
import { BaseVisitor } from './index';
import * as autoBind from 'auto-bind';
import { print, visit } from 'graphql';
import { DepGraph } from 'dependency-graph';
import gqlTag from 'graphql-tag';
import { toPascalCase } from '@graphql-codegen/plugin-helpers';
import { getConfigValue } from './utils';
export class ClientSideBaseVisitor extends BaseVisitor {
    constructor(_fragments, rawConfig, additionalConfig) {
        super(rawConfig, {
            noGraphQLTag: getConfigValue(rawConfig.noGraphQLTag, false),
            gqlImport: rawConfig.gqlImport || null,
            ...additionalConfig,
        });
        this._fragments = _fragments;
        this._collectedOperations = [];
        autoBind(this);
    }
    _getFragmentName(fragment) {
        return (typeof fragment === 'string' ? fragment : fragment.name.value) + 'FragmentDoc';
    }
    _extractFragments(document) {
        if (!document) {
            return [];
        }
        const names = [];
        visit(document, {
            enter: {
                FragmentSpread: (node) => {
                    names.push(node.name.value);
                },
            },
        });
        return names;
    }
    _transformFragments(document) {
        return this._extractFragments(document).map(document => this._getFragmentName(document));
    }
    _includeFragments(fragments) {
        if (fragments && fragments.length > 0) {
            return `${fragments
                .filter((name, i, all) => all.indexOf(name) === i)
                .map(name => '${' + name + '}')
                .join('\n')}`;
        }
        return '';
    }
    _prepareDocument(documentStr) {
        return documentStr;
    }
    _gql(node) {
        const doc = this._prepareDocument(`
    ${print(node)}
    ${this._includeFragments(this._transformFragments(node))}`);
        if (this.config.noGraphQLTag) {
            const gqlObj = gqlTag(doc);
            if (gqlObj && gqlObj['loc']) {
                delete gqlObj.loc;
            }
            return JSON.stringify(gqlObj);
        }
        return 'gql`' + doc + '`';
    }
    _generateFragment(fragmentDocument) {
        const name = this._getFragmentName(fragmentDocument);
        return `export const ${name}${this.config.noGraphQLTag ? ': DocumentNode' : ''} = ${this._gql(fragmentDocument)};`;
    }
    get fragmentsGraph() {
        const graph = new DepGraph({ circular: true });
        for (const fragment of this._fragments) {
            if (graph.hasNode(fragment.name)) {
                const cachedAsString = print(graph.getNodeData(fragment.name).node);
                const asString = print(fragment.node);
                if (cachedAsString !== asString) {
                    throw new Error(`Duplicated fragment called '${fragment.name}'!`);
                }
            }
            graph.addNode(fragment.name, fragment);
        }
        this._fragments.forEach(fragment => {
            const depends = this._extractFragments(fragment.node);
            if (depends && depends.length > 0) {
                depends.forEach(name => {
                    graph.addDependency(fragment.name, name);
                });
            }
        });
        return graph;
    }
    get fragments() {
        if (this._fragments.length === 0) {
            return '';
        }
        const graph = this.fragmentsGraph;
        const orderedDeps = graph.overallOrder();
        const localFragments = orderedDeps.filter(name => !graph.getNodeData(name).isExternal).map(name => this._generateFragment(graph.getNodeData(name).node));
        return localFragments.join('\n');
    }
    _parseImport(importStr) {
        const [moduleName, propName] = importStr.split('#');
        return {
            moduleName,
            propName,
        };
    }
    getImports() {
        const gqlImport = this._parseImport(this.config.gqlImport || 'graphql-tag');
        let imports = [];
        if (!this.config.noGraphQLTag) {
            imports.push(`import ${gqlImport.propName ? `{ ${gqlImport.propName === 'gql' ? 'gql' : `${gqlImport.propName} as gql`} }` : 'gql'} from '${gqlImport.moduleName}';`);
        }
        else {
            imports.push(`import { DocumentNode } from 'graphql';`);
        }
        (this._fragments || [])
            .filter(f => f.isExternal && f.importFrom)
            .forEach(externalFragment => {
            const identifierName = this._getFragmentName(externalFragment.name);
            imports.push(`import { ${identifierName} } from '${externalFragment.importFrom}';`);
        });
        return imports;
    }
    buildOperation(node, documentVariableName, operationType, operationResultType, operationVariablesTypes) {
        return null;
    }
    OperationDefinition(node) {
        if (!node.name || !node.name.value) {
            return null;
        }
        this._collectedOperations.push(node);
        const documentVariableName = this.convertName(node, {
            suffix: 'Document',
            useTypesPrefix: false,
        });
        const documentString = `export const ${documentVariableName}${this.config.noGraphQLTag ? ': DocumentNode' : ''} = ${this._gql(node)};`;
        const operationType = toPascalCase(node.operation);
        const operationResultType = this.convertName(node, {
            suffix: operationType,
        });
        const operationVariablesTypes = this.convertName(node, {
            suffix: operationType + 'Variables',
        });
        const additional = this.buildOperation(node, documentVariableName, operationType, operationResultType, operationVariablesTypes);
        return [documentString, additional].filter(a => a).join('\n');
    }
}
