// @graphql-codegen/typescript-react-apollo/dist/esnext/visitor.js
import { ClientSideBaseVisitor, getConfigValue, OMIT_TYPE } from '@graphql-codegen/visitor-plugin-common';
import * as autoBind from 'auto-bind';
import { Kind } from 'graphql';
import { toPascalCase } from '@graphql-codegen/plugin-helpers';
import { titleCase } from 'change-case';
export class ReactApolloVisitor extends ClientSideBaseVisitor {
    constructor(fragments, rawConfig) {
        super(fragments, rawConfig, {
            componentSuffix: getConfigValue(rawConfig.componentSuffix, 'Component'),
            withHOC: getConfigValue(rawConfig.withHOC, true),
            withComponent: getConfigValue(rawConfig.withComponent, true),
            withHooks: getConfigValue(rawConfig.withHooks, false),
            withMutationFn: getConfigValue(rawConfig.withMutationFn, true),
            hooksImportFrom: getConfigValue(rawConfig.hooksImportFrom, 'react-apollo-hooks'),
            reactApolloImportFrom: getConfigValue(rawConfig.reactApolloImportFrom, 'react-apollo'),
        });
        this.imports = new Set();
        autoBind(this);
    }
    getReactImport() {
        return `import * as React from 'react';`;
    }
    getReactApolloImport() {
        return `import * as ReactApollo from '${typeof this.config.reactApolloImportFrom === 'string' ? this.config.reactApolloImportFrom : 'react-apollo'}';`;
    }
    getReactApolloHooksImport() {
        return `import * as ReactApolloHooks from '${typeof this.config.hooksImportFrom === 'string' ? this.config.hooksImportFrom : 'react-apollo-hooks'}';`;
    }
    getOmitDeclaration() {
        return OMIT_TYPE;
    }
    getImports() {
        const baseImports = super.getImports();
        const hasOperations = this._collectedOperations.length > 0;
        if (!hasOperations) {
            return baseImports;
        }
        return [...baseImports, ...this.imports];
    }
    _buildHocProps(operationName, operationType) {
        const typeVariableName = this.convertName(operationName + toPascalCase(operationType));
        const variablesVarName = this.convertName(operationName + toPascalCase(operationType) + 'Variables');
        const argType = operationType === 'mutation' ? 'MutateProps' : 'DataProps';
        this.imports.add(this.getReactApolloImport());
        return `Partial<ReactApollo.${argType}<${typeVariableName}, ${variablesVarName}>>`;
    }
    _buildMutationFn(node, operationResultType, operationVariablesTypes) {
        if (node.operation === 'mutation') {
            this.imports.add(this.getReactApolloImport());
            return `export type ${this.convertName(node.name.value + 'MutationFn')} = ReactApollo.MutationFn<${operationResultType}, ${operationVariablesTypes}>;`;
        }
        return null;
    }
    _buildOperationHoc(node, documentVariableName, operationResultType, operationVariablesTypes) {
        const operationName = this.convertName(node.name.value, { useTypesPrefix: false });
        const propsTypeName = this.convertName(node.name.value, { suffix: 'Props' });
        const propsVar = `export type ${propsTypeName}<TChildProps = {}> = ${this._buildHocProps(node.name.value, node.operation)} & TChildProps;`;
        const hocString = `export function with${operationName}<TProps, TChildProps = {}>(operationOptions?: ReactApollo.OperationOption<
  TProps,
  ${operationResultType},
  ${operationVariablesTypes},
  ${propsTypeName}<TChildProps>>) {
    return ReactApollo.with${titleCase(node.operation)}<TProps, ${operationResultType}, ${operationVariablesTypes}, ${propsTypeName}<TChildProps>>(${documentVariableName}, {
      alias: 'with${operationName}',
      ...operationOptions
    });
};`;
        return [propsVar, hocString].filter(a => a).join('\n');
    }
    _buildComponent(node, documentVariableName, operationType, operationResultType, operationVariablesTypes) {
        const componentPropsName = this.convertName(node.name.value, { suffix: this.config.componentSuffix + 'Props', useTypesPrefix: false });
        const componentName = this.convertName(node.name.value, { suffix: this.config.componentSuffix, useTypesPrefix: false });
        const isVariablesRequired = operationType === 'Query' && node.variableDefinitions.some(variableDef => variableDef.type.kind === Kind.NON_NULL_TYPE);
        this.imports.add(this.getReactImport());
        this.imports.add(this.getReactApolloImport());
        this.imports.add(this.getOmitDeclaration());
        const componentProps = `export type ${componentPropsName} = Omit<Omit<ReactApollo.${operationType}Props<${operationResultType}, ${operationVariablesTypes}>, '${operationType.toLowerCase()}'>, 'variables'> & { variables${isVariablesRequired ? '' : '?'}: ${operationVariablesTypes} };`;
        const component = `
    export const ${componentName} = (props: ${componentPropsName}) => (
      <ReactApollo.${operationType}<${operationResultType}, ${operationVariablesTypes}> ${node.operation}={${documentVariableName}} {...props} />
    );
    `;
        return [componentProps, component].join('\n');
    }
    _buildHooks(node, operationType, documentVariableName, operationResultType, operationVariablesTypes) {
        const operationName = this.convertName(node.name.value, { suffix: titleCase(operationType), useTypesPrefix: false });
        this.imports.add(this.getReactApolloHooksImport());
        return `
export function use${operationName}(baseOptions?: ReactApolloHooks.${operationType}HookOptions<${node.operation !== 'query' ? `${operationResultType}, ` : ''}${operationVariablesTypes}>) {
  return ReactApolloHooks.use${operationType}<${operationResultType}, ${operationVariablesTypes}>(${documentVariableName}, baseOptions);
};`;
    }
    buildOperation(node, documentVariableName, operationType, operationResultType, operationVariablesTypes) {
        const mutationFn = this.config.withMutationFn || this.config.withComponent ? this._buildMutationFn(node, operationResultType, operationVariablesTypes) : null;
        const component = this.config.withComponent ? this._buildComponent(node, documentVariableName, operationType, operationResultType, operationVariablesTypes) : null;
        const hoc = this.config.withHOC ? this._buildOperationHoc(node, documentVariableName, operationResultType, operationVariablesTypes) : null;
        const hooks = this.config.withHooks ? this._buildHooks(node, operationType, documentVariableName, operationResultType, operationVariablesTypes) : null;
        return [mutationFn, component, hoc, hooks].filter(a => a).join('\n');
    }
}
