// @graphql-codegen/typescript-compatibility/dist/commonjs/visitor.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const visitor_plugin_common_1 = require("@graphql-codegen/visitor-plugin-common");
const selection_set_to_types_1 = require("./selection-set-to-types");
class CompatabilityPluginVisitor extends visitor_plugin_common_1.BaseVisitor {
    constructor(rawConfig, _schema, options) {
        super(rawConfig, {
            reactApollo: options.reactApollo,
            noNamespaces: visitor_plugin_common_1.getConfigValue(rawConfig.noNamespaces, false),
            strict: visitor_plugin_common_1.getConfigValue(rawConfig.strict, false),
        });
        this._schema = _schema;
    }
    getRootType(operationType) {
        if (operationType === 'query') {
            return this._schema.getQueryType().name;
        }
        else if (operationType === 'mutation') {
            return this._schema.getMutationType().name;
        }
        else if (operationType === 'subscription') {
            return this._schema.getSubscriptionType().name;
        }
        return null;
    }
    buildOperationBlock(node) {
        const typeName = this.getRootType(node.operation);
        const baseName = this.convertName(node.name.value, { suffix: `${visitor_plugin_common_1.toPascalCase(node.operation)}` });
        const typesPrefix = this.config.noNamespaces ? this.convertName(node.name.value) : '';
        const selectionSetTypes = {
            [typesPrefix + this.convertName('Variables')]: {
                export: 'type',
                name: this.convertName(node.name.value, { suffix: `${visitor_plugin_common_1.toPascalCase(node.operation)}Variables` }),
            },
        };
        selection_set_to_types_1.selectionSetToTypes(typesPrefix, this, this._schema, typeName, baseName, node.operation, node.selectionSet, selectionSetTypes);
        return selectionSetTypes;
    }
    buildFragmentBlock(node) {
        const typeName = this._schema.getType(node.typeCondition.name.value).name;
        const baseName = this.convertName(node.name.value, { suffix: `Fragment` });
        const typesPrefix = this.config.noNamespaces ? this.convertName(node.name.value) : '';
        const selectionSetTypes = {};
        selection_set_to_types_1.selectionSetToTypes(typesPrefix, this, this._schema, typeName, baseName, 'fragment', node.selectionSet, selectionSetTypes);
        return selectionSetTypes;
    }
    printTypes(selectionSetTypes) {
        return Object.keys(selectionSetTypes)
            .filter(typeName => typeName !== selectionSetTypes[typeName].name)
            .map(typeName => `export ${selectionSetTypes[typeName].export} ${typeName} = ${selectionSetTypes[typeName].name};`)
            .map(m => (this.config.noNamespaces ? m : visitor_plugin_common_1.indent(m)))
            .join('\n');
    }
    FragmentDefinition(node) {
        const baseName = node.name.value;
        const results = [];
        const convertedName = this.convertName(baseName);
        const selectionSetTypes = this.buildFragmentBlock(node);
        const fragmentBlock = this.printTypes(selectionSetTypes);
        if (!this.config.noNamespaces) {
            results.push(new visitor_plugin_common_1.DeclarationBlock(this._declarationBlockConfig)
                .export()
                .asKind('namespace')
                .withName(convertedName)
                .withBlock(fragmentBlock).string);
        }
        else {
            results.push(fragmentBlock);
        }
        return results.join('\n');
    }
    OperationDefinition(node) {
        const baseName = node.name.value;
        const convertedName = this.convertName(baseName);
        const results = [];
        const selectionSetTypes = this.buildOperationBlock(node);
        if (this.config.reactApollo) {
            const reactApolloConfig = this.config.reactApollo[Object.keys(this.config.reactApollo)[0]];
            let hoc = true;
            let component = true;
            let hooks = false;
            if (typeof reactApolloConfig === 'object') {
                if (reactApolloConfig.withHOC === false) {
                    hoc = false;
                }
                if (reactApolloConfig.withComponent === false) {
                    component = false;
                }
                if (reactApolloConfig.withHooks) {
                    hooks = true;
                }
            }
            const prefix = this.config.noNamespaces ? convertedName : '';
            selectionSetTypes[prefix + 'Document'] = {
                export: 'const',
                name: this.convertName(baseName, { suffix: 'Document' }),
            };
            if (hoc) {
                selectionSetTypes[prefix + 'Props'] = {
                    export: 'type',
                    name: this.convertName(baseName, { suffix: 'Props' }),
                };
                selectionSetTypes[prefix + 'HOC'] = {
                    export: 'const',
                    name: `with${convertedName}`,
                };
            }
            if (component) {
                selectionSetTypes[prefix + 'Component'] = {
                    export: 'const',
                    name: this.convertName(baseName, { suffix: 'Component' }),
                };
            }
            if (hooks) {
                selectionSetTypes['use' + prefix] = {
                    export: 'const',
                    name: 'use' + this.convertName(baseName, { suffix: visitor_plugin_common_1.toPascalCase(node.operation) }),
                };
            }
        }
        const operationsBlock = this.printTypes(selectionSetTypes);
        if (!this.config.noNamespaces) {
            results.push(new visitor_plugin_common_1.DeclarationBlock(this._declarationBlockConfig)
                .export()
                .asKind('namespace')
                .withName(convertedName)
                .withBlock(operationsBlock).string);
        }
        else {
            results.push(operationsBlock);
        }
        return results.join('\n');
    }
}
exports.CompatabilityPluginVisitor = CompatabilityPluginVisitor;
