// @graphql-codegen/core/dist/commonjs/codegen.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const plugin_helpers_1 = require("@graphql-codegen/plugin-helpers");
const graphql_1 = require("graphql");
const merge_schemas_1 = require("./merge-schemas");
const execute_plugin_1 = require("./execute-plugin");
const errors_1 = require("./errors");
async function codegen(options) {
    let output = '';
    validateDocuments(options.schema, options.documents);
    const pluginPackages = Object.keys(options.pluginMap).map(key => options.pluginMap[key]);
    // merged schema with parts added by plugins
    const schema = pluginPackages.reduce((schema, plugin) => {
        return !plugin.addToSchema ? schema : merge_schemas_1.mergeSchemas([schema, plugin.addToSchema]);
    }, options.schema);
    const prepend = new Set();
    const append = new Set();
    for (const plugin of options.plugins) {
        const name = Object.keys(plugin)[0];
        const pluginPackage = options.pluginMap[name];
        const pluginConfig = plugin[name];
        const result = await execute_plugin_1.executePlugin({
            name,
            config: typeof pluginConfig !== 'object'
                ? pluginConfig
                : {
                    ...options.config,
                    ...pluginConfig,
                },
            schema,
            documents: options.documents,
            outputFilename: options.filename,
            allPlugins: options.plugins,
        }, pluginPackage);
        if (typeof result === 'string') {
            output += result;
        }
        else if (plugin_helpers_1.isComplexPluginOutput(result)) {
            output += result.content || '';
            if (result.append && result.append.length > 0) {
                for (const item of result.append) {
                    append.add(item);
                }
            }
            if (result.prepend && result.prepend.length > 0) {
                for (const item of result.prepend) {
                    prepend.add(item);
                }
            }
        }
    }
    return [...sortPrependValues(Array.from(prepend.values())), output, ...append.values()].join('\n');
}
exports.codegen = codegen;
function resolveCompareValue(a) {
    if (a.startsWith('/*') || a.startsWith('//')) {
        return 0;
    }
    else if (a.startsWith('import')) {
        return 1;
    }
    else {
        return 2;
    }
}
function sortPrependValues(values) {
    return values.sort((a, b) => {
        const aV = resolveCompareValue(a);
        const bV = resolveCompareValue(b);
        if (aV < bV) {
            return -1;
        }
        if (aV > bV) {
            return 1;
        }
        return 0;
    });
}
exports.sortPrependValues = sortPrependValues;
function validateDocuments(schema, files) {
    // duplicated names
    const operationMap = {};
    files.forEach(file => {
        graphql_1.visit(file.content, {
            OperationDefinition(node) {
                if (typeof node.name !== 'undefined') {
                    if (!operationMap[node.name.value]) {
                        operationMap[node.name.value] = [];
                    }
                    operationMap[node.name.value].push(file.filePath);
                }
            },
        });
    });
    const names = Object.keys(operationMap);
    if (names.length) {
        const duplicated = names.filter(name => operationMap[name].length > 1);
        if (!duplicated.length) {
            return;
        }
        const list = duplicated
            .map(name => `
      * ${name} found in:
        ${operationMap[name]
            .map(filepath => {
            return `
            - ${filepath}
          `.trimRight();
        })
            .join('')}
  `.trimRight())
            .join('');
        throw new errors_1.DetailedError(`Not all operations have an unique name: ${duplicated.join(', ')}`, `
        Not all operations have an unique name

        ${list}
      `);
    }
}
