// graphql-toolkit/dist/commonjs/sonar/file-scanner.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const glob_1 = require("glob");
const glob = require("glob");
const path_1 = require("path");
const fs_1 = require("fs");
const graphql_1 = require("graphql");
const DEFAULT_SCHEMA_EXTENSIONS = ['gql', 'graphql', 'graphqls', 'ts', 'js'];
const DEFAULT_IGNORED_RESOLVERS_EXTENSIONS = ['spec', 'test', 'd'];
const DEFAULT_RESOLVERS_EXTENSIONS = ['ts', 'js'];
const DEFAULT_SCHEMA_EXPORT_NAMES = ['typeDefs', 'schema'];
const DEFAULT_RESOLVERS_EXPORT_NAMES = ['resolvers', 'resolver'];
function scanForFiles(globStr, globOptions = {}) {
    return glob_1.sync(globStr, { absolute: true, ...globOptions });
}
function buildGlob(basePath, extensions, ignoredExtensions = []) {
    return `${basePath}/**/${ignoredExtensions.length > 0 ? `!(${ignoredExtensions.map(e => '*.' + e).join('|')})` : '*'}+(${extensions.map(e => '*.' + e).join('|')})`;
}
function extractExports(fileExport, exportNames) {
    if (!fileExport) {
        return null;
    }
    if (fileExport.default) {
        for (const exportName of exportNames) {
            if (fileExport.default[exportName]) {
                return fileExport.default[exportName];
            }
        }
        return fileExport.default;
    }
    for (const exportName of exportNames) {
        if (fileExport[exportName]) {
            return fileExport[exportName];
        }
    }
    return fileExport;
}
const LoadSchemaFilesDefaultOptions = {
    extensions: DEFAULT_SCHEMA_EXTENSIONS,
    useRequire: false,
    requireMethod: null,
    globOptions: {},
    exportNames: DEFAULT_SCHEMA_EXPORT_NAMES,
};
function loadSchemaFiles(basePath, options = LoadSchemaFilesDefaultOptions) {
    const execOptions = { ...LoadSchemaFilesDefaultOptions, ...options };
    const relevantPaths = scanForFiles(buildGlob(basePath, execOptions.extensions, []), options.globOptions);
    return relevantPaths.map(path => {
        const extension = path_1.extname(path);
        if (extension === '.js' || extension === '.ts' || execOptions.useRequire) {
            const fileExports = (execOptions.requireMethod ? execOptions.requireMethod : require)(path);
            const extractedExport = extractExports(fileExports, execOptions.exportNames);
            if (extractedExport && extractedExport.kind === 'Document') {
                return graphql_1.print(extractedExport);
            }
            return extractedExport;
        }
        else {
            return fs_1.readFileSync(path, { encoding: 'utf-8' });
        }
    });
}
exports.loadSchemaFiles = loadSchemaFiles;
const LoadResolversFilesDefaultOptions = {
    ignoredExtensions: DEFAULT_IGNORED_RESOLVERS_EXTENSIONS,
    extensions: DEFAULT_RESOLVERS_EXTENSIONS,
    requireMethod: null,
    globOptions: {},
    exportNames: DEFAULT_RESOLVERS_EXPORT_NAMES,
};
function loadResolversFiles(basePath, options = LoadResolversFilesDefaultOptions) {
    const execOptions = { ...LoadResolversFilesDefaultOptions, ...options };
    const relevantPaths = scanForFiles(buildGlob(basePath, execOptions.extensions, execOptions.ignoredExtensions), execOptions.globOptions);
    return relevantPaths.map(path => {
        try {
            const fileExports = (execOptions.requireMethod ? execOptions.requireMethod : require)(path);
            return extractExports(fileExports, execOptions.exportNames);
        }
        catch (e) {
            throw new Error(`Unable to load resolver file: ${path}, error: ${e}`);
        }
    }).filter(t => t);
}
exports.loadResolversFiles = loadResolversFiles;
function scanForFilesAsync(globStr, globOptions = {}) {
    return new Promise((resolve, reject) => glob(globStr, { absolute: true, ...globOptions }, (err, matches) => {
        if (err) {
            reject(err);
        }
        resolve(matches);
    }));
}
async function loadSchemaFilesAsync(basePath, options = LoadSchemaFilesDefaultOptions) {
    const execOptions = { ...LoadSchemaFilesDefaultOptions, ...options };
    const relevantPaths = await scanForFilesAsync(buildGlob(basePath, execOptions.extensions, []), options.globOptions);
    const require$ = (path) => Promise.resolve().then(() => require(path));
    return Promise.all(relevantPaths.map(async (path) => {
        const extension = path_1.extname(path);
        if (extension === '.js' || extension === '.ts' || execOptions.useRequire) {
            const fileExports = await (execOptions.requireMethod ? execOptions.requireMethod : require$)(path);
            const extractedExport = extractExports(fileExports, execOptions.exportNames);
            if (extractedExport && extractedExport.kind === 'Document') {
                return graphql_1.print(extractedExport);
            }
            return extractedExport;
        }
        else {
            return new Promise((resolve, reject) => {
                fs_1.readFile(path, { encoding: 'utf-8' }, (err, data) => {
                    if (err) {
                        reject(err);
                    }
                    resolve(data);
                });
            });
        }
    }));
}
exports.loadSchemaFilesAsync = loadSchemaFilesAsync;
async function loadResolversFilesAsync(basePath, options = LoadResolversFilesDefaultOptions) {
    const execOptions = { ...LoadResolversFilesDefaultOptions, ...options };
    const relevantPaths = await scanForFilesAsync(buildGlob(basePath, execOptions.extensions, []), options.globOptions);
    const require$ = (path) => Promise.resolve().then(() => require(path));
    return Promise.all(relevantPaths.map(async (path) => {
        try {
            const fileExports = await (execOptions.requireMethod ? execOptions.requireMethod : require$)(path);
            return extractExports(fileExports, execOptions.exportNames);
        }
        catch (e) {
            throw new Error(`Unable to load resolver file: ${path}, error: ${e}`);
        }
    }));
}
exports.loadResolversFilesAsync = loadResolversFilesAsync;
